src/mlpack/core/math/random.hpp

```diff
@@ -23,12 +23,42 @@ namespace math /** Miscellaneous math routines. */ {
  * correctly on Windows.
  */
 
-// Global random object.
-extern MLPACK_EXPORT std::mt19937 randGen;
-// Global uniform distribution.
-extern MLPACK_EXPORT std::uniform_real_distribution<> randUniformDist;
-// Global normal distribution.
-extern MLPACK_EXPORT std::normal_distribution<> randNormalDist;
+#ifndef MLPACK_CORE_MATH_RANDOM_GLOBAL
+#define MLPACK_CORE_MATH_RANDOM_GLOBAL
+
+#if __cplusplus < 201703L
+    template<class Dummy>
+    struct GlobalRandomVariables_
+    {
+      // Global random object.
+      static std::mt19937 randGen;
+      // Global uniform distribution.
+      static std::uniform_real_distribution<> randUniformDist;
+      // Global normal distribution.
+      static std::normal_distribution<> randNormalDist;
+    };
+    template<class Dummy>
+    // Global random object.
+    std::mt19937 GlobalRandomVariables_<Dummy>::randGen;
+    template<class Dummy>
+    // Global uniform distribution.
+    std::uniform_real_distribution<> GlobalRandomVariables_<Dummy>::randUniformDist(0.0, 1.0);
+    template<class Dummy>
+    // Global normal distribution.
+    std::normal_distribution<> GlobalRandomVariables_<Dummy>::randNormalDist(0.0, 1.0);
+    using GlobalRandomVariables = GlobalRandomVariables_<void>;
+#else
+  namespace GlobalRandomVariables {
+    // Global random object.
+    inline std::mt19937 randGen;
+    // Global uniform distribution.
+    inline std::uniform_real_distribution<> randUniformDist(0.0, 1.0);
+    // Global normal distribution.
+    inline std::normal_distribution<> randNormalDist(0.0, 1.0);
+  }
+#endif
+
+#endif
```

Comment:
I'm wondering if the change to random behavior here is what is causing the test failures.  Maybe it makes sense to use a function with a static member inside of it?

```
inline std::mt19937& GetRandGen()
{
  static std::mt19937 randGen;
  return randGen;
}
```

then replace all references to `randGen` with `GetRandGen()`?  (Similarly for `randNormalDist` and `randUniformDist`.)  I believe that solution also shouldn't need different behavior for pre- and post-C++17.

Reply:
Which test cases are broken? Can you please specify

Reply:
I'm just looking at the Linux builds on Azure; I see output like this:

```
-------------------------------------------------------------------------------
VanillaNetworkBatchSizeTest
-------------------------------------------------------------------------------
/home/vsts/work/1/s/src/mlpack/tests/convolutional_network_test.cpp:254
...............................................................................

/home/vsts/work/1/s/src/mlpack/tests/test_catch_tools.hpp:37: FAILED:
  REQUIRE( a[i] == Approx(b[i]).epsilon(tolerance / 100) )
with expansion:
  4.6500060959 == Approx( 11.5978194139 )

-------------------------------------------------------------------------------
KFoldCVPerceptronTest
-------------------------------------------------------------------------------
/home/vsts/work/1/s/src/mlpack/tests/cv_test.cpp:619
...............................................................................

/home/vsts/work/1/s/src/mlpack/tests/cv_test.cpp:636: FAILED:
  REQUIRE( cv.Evaluate() == Approx(expectedAccuracy).epsilon(1e-7) )
with expansion:
  1.0 == Approx( 0.9 )
```

For instance, https://dev.azure.com/mlpack/mlpack/_build/results?buildId=8012&view=logs&jobId=9bb2c93b-115c-5be3-b0bf-bdc0e9bf3db4&j=204580d5-7a27-510d-d602-9122b08edba8&t=f116dad7-ff86-58c7-0b07-4fe3b687b187

You could quickly validate the theory that the random changes are the issue simply by reverting the changes to random.hpp and seeing what happens.  Anyway, let me know if I can clarify anything. :+1:

Reply:
I think these are independent of this code as it was showing in other PR too

Reply:
Agreed, sorry I didn't realize that.  I opened #3212 to fix one of them, and I'm still working to reproduce the `KFoldCVPerceptronTest` failure.

Reply:
Nonetheless, what do you think about using a static member inside a static function like I suggested?  I think this is a cleaner design, if it works.

Reply:
> Nonetheless, what do you think about using a static member inside a static function like I suggested? I think this is a cleaner design, if it works.

Using `static std::mt19937 randGen` inside a function could be problematic if the function is called from multiple threads.

At worst there may be data races.  For example, multiple threads may try to initialise the `randGen` variable first.  Even if the `randGen` variable is correctly initialised, using it from multiple threads can unpredictably affect its internal state, potentially leading to crashes.

In the best situation the quasi-random numbers provided by `randGen` would be non-deterministic if called from multiple threads. This may at first sound desirable (ie. the more randomness the better), but this would affect reproducibility, and/or produce numbers that no longer follow the desired distribution.

If using a static function is really desired, I suggest looking at the `thread_local` qualifier:  
* https://en.cppreference.com/w/cpp/language/storage_duration
* https://en.wikipedia.org/wiki/Thread-local_storage

Reply:
Thanks for pointing that out---you are absolutely right. `thread_local` will be needed here.  In fact I think the current RNG is not thread-safe as it stands now, so this will fix a bug. :+1:

Reply:
Can you write this as a suggestion?

Reply:
I can't put it in the right place as a suggestion, but it would be something like this:

```
inline std::mt19937& RandGen()
{
  static thread_local std::mt19937 randGen;
  return randGen;
}
```

I think that is about right, I didn't actually try it yet.  Then every place where you are referencing `GlobalRandomVariables::randGen`, you can replace it with `RandGen()`.

Reply:
@shubham1206agra let me know if I can help out with this---I think this is the only remaining issue to handle before we can merge this. :+1:

Reply:
Thanks for the changes---they look good to me.  But something must not be working right; the tests timed out.

Reply:
I think there is some issue with thread_local

Reply:
This has been fixed now.